river/command/view_operations.zig

```diff
@@ -0,0 +1,133 @@
+//
+// Copyright 2020 - 2023 The River Developers
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const std = @import("std");
+const assert = std.debug.assert;
+const wlr = @import("wlroots");
+
+const server = &@import("../main.zig").server;
+
+const Direction = @import("../command.zig").Direction;
+const Error = @import("../command.zig").Error;
+const Output = @import("../Output.zig");
+const Seat = @import("../Seat.zig");
+const View = @import("../View.zig");
+const Vector = @import("../Vector.zig");
+
+/// Focus either the next or the previous visible view, depending on the enum
+/// passed. Does nothing if there are 1 or 0 views in the stack.
+pub fn focusView(
+    seat: *Seat,
+    args: []const [:0]const u8,
+    _: *?[]const u8,
+) Error!void {
+    if (args.len < 2) return Error.NotEnoughArguments;
+    if (args.len > 2) return Error.TooManyArguments;
+
+    if (try getTarget(seat, args[1], .all)) |target| {
+        assert(!target.pending.fullscreen);
+        seat.focus(target);
+        server.root.applyPending();
+    }
+}
+
+/// Swap the currently focused view with either the view higher or lower in the visible stack
+pub fn swap(
+    seat: *Seat,
+    args: []const [:0]const u8,
+    _: *?[]const u8,
+) Error!void {
+    if (args.len < 2) return Error.NotEnoughArguments;
+    if (args.len > 2) return Error.TooManyArguments;
+
+    if (try getTarget(seat, args[1], .skip_float)) |target| {
+        assert(!target.pending.float);
+        assert(!target.pending.fullscreen);
+        seat.focused.view.pending_wm_stack_link.swapWith(&target.pending_wm_stack_link);
+        server.root.applyPending();
+    }
+}
+
+const TargetMode = enum { all, skip_float };
+fn getTarget(seat: *Seat, direction_str: []const u8, target_mode: TargetMode) !?*View {
+    if (seat.focused != .view) return null;
+    if (seat.focused.view.pending.fullscreen) return null;
+    if (target_mode == .skip_float and seat.focused.view.pending.float) return null;
+    const output = seat.focused_output orelse return null;
+
+    // If no currently view is focused, focus the first in the stack.
+    if (seat.focused != .view) {
+        var it = output.pending.wm_stack.iterator(.forward);
+        return it.next();
+    }
+
+    // Logical direction, based on the view stack.
+    if (std.meta.stringToEnum(Direction, direction_str)) |direction| {
+        switch (direction) {
+            inline else => |dir| {
```

Comment:
Looking over this again, I don't see why there is a switch with an `inline else` here. I copied that from the current implementation, but actually looking at and thinking about this code, the only place `dir` is used is in the inner switch to get `it_dir`, so I'd be genuinely surprised if the outer switch actually has any benefits w.r.t. to machine code generation.

Reply:
Doesn't the `iterator()` function used in this branch take a comptime value?

Reply:
You're right, it does.